### ISSUE TITLE
Clojure compilation of tensor code just worked ~~ (first time)

### ISSUE THREAD

**Issue by cnuernber** [dmlc/tvm#1035]
I managed to compile generic tensor code:

```clojure
(defn convert-bgr-bytes-to-floats
  "Input is unsigned bytes, so values 0->255, bgr [height width 3].
Output: {:datatype :float32 :shape [3 height width]}, values from -0.5->0.5"
  [input-tensor]
  ;;Move to bgr instead of rgb  Also drop alpha if exists in first place.
  (-> (ct/select input-tensor :all :all [2 1 0])
      (ct/transpose [2 0 1])
      ;;First actual operation that is compiled.
      ;;Rest of the work.  These should all get rolled into assignment step above.
      (ct/static-cast :float32)
      (ct/div 255.0)
      (ct/sub 0.5)))


(defn compile-bgr-bytes
  [& {:keys [driver-name]
      :or {driver-name :cpu}}]
  (let [graph (-> (compiler/compute-graph)
                  (compiler/make-variable :image-width)
                  (compiler/make-variable :image-height)
                  (compiler/make-variable :image-channels)
                  (compiler/make-tensor-and-buffer :input [:image-height :image-width :image-channels] :dtype :uint8))
        input-tensor (compiler/get-tensor graph :input)]
    (assoc (compiler/compile-fn (comp-base/get-driver driver-name) graph convert-bgr-bytes-to-floats input-tensor)
           :driver driver-name)))

(defn convert-bgr-bytes-to-floats-by-hand
  [input-tensor result-tensor]
  (let [^bytes ary-data (.data ^ByteArrayView (:buffer input-tensor))
        [height width n-channels] (ct/shape input-tensor)
        height (long height)
        width (long width)
        n-channels (long n-channels)
        n-elems (long (* height width 3))
        ^floats retval (.data ^FloatArrayView (:buffer result-tensor))]
    (parallel/parallel-for idx n-elems
           (let [pixel (quot idx n-channels)
                 channel (rem idx n-channels)
                 x-pos (rem pixel width)
                 y-pos (quot pixel width)
                 dest-channel-stride (* x-pos y-pos)]
             (when (< channel 3)
               (aset retval (+ pixel
                               (* (- 2 channel) dest-channel-stride))
                     (float (- (/ (aget ary-data idx)
                                  255.0)
                               0.5))))))
    result-tensor))
```
```

tvm-clj.compute.compile-fn-test> (time-tests)
java tensor ops took:  "Elapsed time: 7661.258735 msecs"

hand-coded java took:  "Elapsed time: 557.899507 msecs"

Compiled tensor took: "Elapsed time: 342.164241 msecs"

nil
```

I ran it on an image of dimensions: 2988x5312x3 

https://github.com/tech-ascent/tvm-clj/blob/master/test/tvm_clj/compute/compile_fn_test.clj

This is huge for a lot of reasons.  One interesting thing to note is that the java versions produce incorrect results due to the fact that the jvm doesn't support unsigned datatypes.  The tvm version produces the correct results (at least for the first 10 numbers; I checked).

I used a [custom indexing scheme](https://github.com/tech-ascent/tvm-clj/blob/master/src/tvm_clj/compute/compile_fn.clj#L505) for the input tensor so I could support a compiled version of the bgr translation.

Thoughts?

**Comment by tqchen** [comment]
was traveling in last few days, this is great! Once you feel it is good to reach out to folks, maybe you could send a guest blogpost to tvmlang.org . I am less familiar with what clojure users care about but maybe you have better ideas of motivating examples

**Comment by cnuernber** [comment]
I am also less familiar with clojure users than one might think.  I like it and am very productive with it so that is good enough for me.  TVM is a *great* counterpart to clojure regardless as the Venn diagrams of their respective strengths do not overlap. 


I sprinted to get the above to work; it is really more of a vision of what could be than a reality.  I only support injective kernels and I did the kernel combining manually as opposed to using the scheduling system to do it.  The TOPI library has a *lot* of very good complex pieces that I want to comb through and the other articles on the tvmlang.org page contain quite a lot of valuable information.
  

So, thanks, I will start working on what will probably be a series of blogposts because I am targeting at least 2 non-overlapping audiences; the core talent behind clojure and the core talent behind tvm.
